### PR TITLE
feat(combobox): enable custom rendering for basic combobox

### PR DESCRIPTION
**Pull request by anitavincent**
#### What is the purpose of this pull request?
Add support for object lists on combobox and add support to custom rendering on the options.

#### What problem is this solving?
Feature requested

#### How should this be manually tested?
built storybook

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Quality improvement (tests or refactors)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Trivial change (small fix or feature that doesn't impact functionalities)
- [ ] Requires change to documentation, which has been updated accordingly

### CODE REVIEW

File: packages/admin-ui/src/combobox/combobox-item.ts [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
-export const ComboboxItem = createComponent<typeof AriakitComboboxItem>(
+export const ComboboxItem = createComponent<'div', ComboboxItemOptions>(
   (props) => {
-    return useElement(AriakitComboboxItem, {
-      ...props,
+    // weird, the input of the hook should be exactly ComboboxItemOptions
+    const htmlProps = useComboboxItem(props as any) as Omit<
+      HTMLProps<'div'>,
+      never
+    >
```
[the remark this answers is not in the document]
Comment (matheusps) [reply packages/admin-ui/src/combobox/combobox-item.ts:14]: But I'm not sure why you switched the approach here. Any specific reason?

File: packages/admin-ui/src/combobox/combobox-popover-base.tsx [hunk 2]
```diff
@@ -107,13 +107,13 @@ export const ComboboxPopoverBase = createComponent<
 
   return useElement(AriakitComboboxPopover, {
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/combobox-popover-base.tsx:108]: Why not switch the approach here?

File: packages/admin-ui/src/combobox/combobox-popover-base.tsx [hunk 3]
```diff
@@ -107,13 +107,13 @@ export const ComboboxPopoverBase = createComponent<
...
-    state: comboboxState,
+    state: { ...comboboxState, matches: [] },
     baseStyle: style.popover,
     children: renderChildren(),
   })
 })
 
 interface Props {
-  state: ComboboxState
+  state: ComboboxState<any>
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/combobox-popover-base.tsx:117]: Same about generics

File: packages/admin-ui/src/combobox/combobox.state.ts [hunk 4]
```diff
@@ -33,24 +76,58 @@ export function useComboboxState(props: ComboboxStateProps = {}) {
...
+  /** Function for transforming item shape into a string value, no need to use it on string[] lists */
+  getOptionValue?: (item: T) => string
+  /** Function for transforming item shape into renderable node, no need to use it on string[] lists */
+  renderOption?: (item: T) => ReactNode
+  /** Debounce interval */
   timeoutMs?: number
 }
 
-export type ComboboxState = ReturnType<typeof useComboboxState>
+export type ComboboxState<T> = Omit<AriakitComboboxState, 'matches'> & {
```
Comment (matheusps) [inline packages/admin-ui/src/combobox/combobox.state.ts:112]: What's wrong with the `ReturnType`?

File: packages/admin-ui/src/combobox/combobox-item.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (anitavincent) [reply packages/admin-ui/src/combobox/combobox-item.ts:14]: Ariakit typing bugs that were breaking the build

[the remark this answers is not in the document]
Comment (matheusps) [reply packages/admin-ui/src/combobox/combobox-item.ts:14]: Makes sense...Should we change the rest then?

[the remark this answers is not in the document]
Comment (anitavincent) [reply packages/admin-ui/src/combobox/combobox-item.ts:14]: The other ones aren't causing errors for now, in my opinion this one should be changed in the future... The issue was that `AriakitComboboxItem` wasn't accepting the props children and onClick, when I think those are actually supposed to be used

File: packages/admin-ui/src/combobox/combobox.state.ts (hunk 4 above)
Comment (anitavincent) [reply packages/admin-ui/src/combobox/combobox.state.ts:112]: I needed to use the generic type on `matches: T[]`

File: packages/admin-ui/src/combobox/combobox-item.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (matheusps) [reply packages/admin-ui/src/combobox/combobox-item.ts:14]: Got it

File: packages/admin-ui/src/combobox/combobox.state.ts (hunk 4 above)
Comment (matheusps) [reply packages/admin-ui/src/combobox/combobox.state.ts:112]: Makes sense

**Review by matheusps — COMMENTED** [review]
The only thing that is missing for me, is some example in the docs on how to do that. Besides that, all clear

**Comment by lucasaarcoverde** [comment]
Done in another pr https://github.com/vtex/admin-ui/pull/122

**Pull request closed without merging**